File: packages/server/src/slidra/history-group.ts
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Copyright contributors to the Slidra project

import { SlidraError } from "./errors.js";
import { runJsonCommand } from "./command.js";

/**
 * `beginHistoryGroup`/`endHistoryGroup` — [E6.T6]: undo history moved
 * inside the deck file itself (three side tables next to `content`,
 * `crates/slidra/src/history.rs`'s module doc), so `packages/server` no
 * longer reads/writes `stack.json`'s `openGroup` field directly. These two
 * functions are now thin wrappers over the Rust binary's own internal
 * `slidra history begin-group`/`slidra history end-group` commands
 * (`crates/slidra/src/commands/history_group.rs`, `docs/spec/cli.md`'s
 * "## Internal commands" — not part of the agent-facing command surface),
 * which themselves wrap `crate::history::begin_history_group`/
 * `end_history_group`.
 *
 * `crates/slidra/src/history.rs`'s own comment documents the exact
 * co-existence this preserves: the Rust binary's own commands append onto
 * whatever group this module leaves open, and close it themselves only
 * when there is none — the two sides are still reading and writing the
 * *same* deck's open-group row, not two independent stores. The one-
 * history-group-per-turn contract `agent/session.ts` relies on is
 * unchanged; only the storage these two calls go through has moved.
 */

/**
 * Opens a group that spans multiple commands (an agent's turn) so they undo
 * together as one step. Returns `true` when this call is the one that
 * opened the group — the caller owns it and MUST call `endHistoryGroup` in
 * a `finally` block. Returns `false` when a group was already open — the
 * caller has joined it and MUST NOT close it.
 */
export async function beginHistoryGroup(id: string): Promise<boolean> {
  const result = await runJsonCommand<{ opened: boolean }>(["history", "begin-group", id]);
  if (!result.ok) {
    throw new SlidraError(result.message);
  }
  return result.data?.opened ?? false;
}

/**
 * Closes the group opened by `beginHistoryGroup` and pushes it onto the
 * undo stack as one step. An empty group (no command in it ever wrote
 * anything) is discarded rather than pushed.
 */
export async function endHistoryGroup(id: string): Promise<void> {
  const result = await runJsonCommand(["history", "end-group", id]);
  if (!result.ok) {
    throw new SlidraError(result.message);
  }
}

File: crates/slidra/src/commands/history_group.rs
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Copyright contributors to the Slidra project

//! `slidra history begin-group <presentation-id>` / `history end-group
//! <presentation-id>` — [E6.T6] D4: internal-only commands, never exposed
//! via `docs/spec/cli.md`'s `` ## `name` `` heading format (see that
//! document's own "## Internal commands" section). The sole caller is
//! `packages/server`'s own turn grouping (`agent/session.ts`), which used
//! to read/write `stack.json`'s `openGroup` field directly and now calls
//! these two commands instead. Thin wrappers over
//! `history::begin_history_group`/`history::end_history_group`.

use crate::history;
use crate::result::{CommandResult, FailureKind};

pub fn dispatch(tokens: super::CommandTokens, args: &[String]) -> CommandResult {
    match tokens[1] {
        "begin-group" => begin_group(args),
        "end-group" => end_group(args),
        _ => unreachable!("resolve_takeover only ever returns entries from this table"),
    }
}

fn begin_group(args: &[String]) -> CommandResult {
    let Some(id) = args.first() else {
        return CommandResult::failure(
            "command history begin-group missing argument: presentation-id",
            FailureKind::Failed,
        );
    };
    match history::begin_history_group(id) {
        Ok(opened) => {
            let message = if opened {
                "opened a new undo group"
            } else {
                "joined an already-open undo group"
            };
            CommandResult::success(message, Some(serde_json::json!({ "opened": opened })))
        }
        Err(err) => CommandResult::from_error(&err),
    }
}

fn end_group(args: &[String]) -> CommandResult {
    let Some(id) = args.first() else {
        return CommandResult::failure(
            "command history end-group missing argument: presentation-id",
            FailureKind::Failed,
        );
    };
    match history::end_history_group(id) {
        Ok(()) => CommandResult::success("closed the undo group", None),
        Err(err) => CommandResult::from_error(&err),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn begin_group_missing_id_argument_fails_without_falling_back_to_node() {
        let result = begin_group(&[]);
        assert!(!result.ok);
        assert_eq!(
            result.message,
            "command history begin-group missing argument: presentation-id"
        );
    }

    #[test]
    fn end_group_missing_id_argument_fails_without_falling_back_to_node() {
        let result = end_group(&[]);
        assert!(!result.ok);
        assert_eq!(
            result.message,
            "command history end-group missing argument: presentation-id"
        );
    }

    #[test]
    fn begin_group_unknown_id_is_not_found() {
        let result = begin_group(&["definitely-not-a-real-id-in-any-test-fixture".to_string()]);
        assert!(!result.ok);
        assert_eq!(result.failure_kind, Some(FailureKind::NotFound));
    }

    #[test]
    fn end_group_unknown_id_is_not_found() {
        let result = end_group(&["definitely-not-a-real-id-in-any-test-fixture".to_string()]);
        assert!(!result.ok);
        assert_eq!(result.failure_kind, Some(FailureKind::NotFound));
    }
}
